cmd/clusterctl/pkg/client/repository/repository_local_unix.go

```diff
@@ -0,0 +1,240 @@
+		}
+		return content, nil
+	}
+	return nil, errors.Errorf("failed to read files from local release %s, path %s", version, absolutePath)
+
+}
+
+// GetVersion returns the list of versions that are available for a local repository
+func (r *localRepository) GetVersions() ([]string, error) {
+	versions, err := r.getVersions()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get local repository versions")
+	}
+	return versions, nil
+}
+
+// newLocalRepository returns a localRepository implementation
+func newLocalRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient) (*localRepository, error) {
+
+	var err error
+
+	urlSplit, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+	absPath := urlSplit.Path
+	if urlSplit.RawPath != "" {
+		absPath = urlSplit.RawPath
+	}
+
+	if !filepath.IsAbs(absPath) {
+		return nil, errors.Errorf("invalid path: path %q must be an absolute path", providerConfig.URL())
+	}
+
+	parts := strings.Split(providerConfig.URL(), "/")
+	// {basepath}/{provider-name}/{version}/{components.yaml}
+	if len(parts) < 3 {
+		return nil, errors.Errorf("invalid path: path should be in the form {basepath}/{provider-name}/{version}/{components.yaml}")
+	}
+	// We work our way backwards with {components.yaml} being the last part of the path
+	componentsPath := parts[len(parts)-1]
+	defaultVersion := parts[len(parts)-2]
+	if defaultVersion != "latest" {
+		_, err = version.ParseSemantic(defaultVersion)
+		if err != nil {
+			return nil, errors.Errorf("invalid version: %q. Version must obey the syntax and semantics of the \"Semantic Versioning\" specification (http://semver.org/) and path format {basepath}/{provider-name}/{version}/{components.yaml}", defaultVersion)
+		}
+	}
+	providerName := parts[len(parts)-3]
+	if providerName != providerConfig.Name() {
+		return nil, errors.Errorf("invalid path: path %q must contain provider name %q in the format {basepath}/{provider-name}/{version}/{components.yaml}", providerConfig.URL(), providerConfig.Name())
+	}
+	var basePath string
+	if len(parts) > 3 {
+		basePath = filepath.Join(parts[:len(parts)-3]...)
+	}
+	basePath = filepath.Clean("/" + basePath) // ensure basePath starts with "/"
+
+	repo := &localRepository{
+		providerConfig:        providerConfig,
+		configVariablesClient: configVariablesClient,
+		basepath:              basePath,
+		providerName:          providerName,
+		defaultVersion:        defaultVersion,
+		rootPath:              "", // Not applicable with localRepository
+		componentsPath:        componentsPath,
+	}
+
+	if defaultVersion == "latest" {
+		repo.defaultVersion, err = repo.getLatestRelease()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get latest version")
+		}
+	}
+
+	return repo, nil
+}
+
+// getLatestRelease returns the latest release for the local repository.
+func (r *localRepository) getLatestRelease() (string, error) {
```

Comment:
What about rewriting this func using the output of getVersions instead of files? I think the resulting code will be simpler

Reply:
good point. done!